project-addons/purchase_proposal/product.py

```diff
@@ -20,12 +20,28 @@
 ##############################################################################
 from openerp import models, fields, api
 import datetime
+import openerp.addons.decimal_precision as dp
 
 
 class ProductProduct(models.Model):
 
     _inherit = 'product.product'
 
+    @api.one
+    @api.depends('standard_price')
+    @api.onchange('standard_price')
+    def _get_pvm(self):
```

Comment:
Esto se podría hacer de una forma mejor, ya que Odoo tiene funciones que nos dan el precio de una tarifa para un producto así que se podría hacer uso de este y no depenedes de que la tarifa este'configrada de una forma, ya que en tu función si la tarifa tuviese más de una versión o más de una regla no devolvería el resultado correctamente:
```
@api.multi
@api.depends('standard_price')
@api.onchange('standard_price')
def _get_pvm(self):
    pricelist = self.env['product.pricelist'].search_read([('name', '=', 'PVM')], ['id'])
    new_ctx = dict(context)
    new_ctx['pricelist'] = pricelist[0]['id']
    for product in self.with_context(new_ctx): # esto no se si funciona sino habría que volver a instanciar los self.ids con el contexto en un browse o usar api.one pero así nos evitamos que en una entrada múltiple haya que hacer el search_read en tarifas en cada entrada.
        product.pvm_price = product.price # este campo es calculado y nos devuelve el precio del prodiucto según la tarifa en contexto, acepta también otros parámetros
```

Reply:
Con esto no me vienen los cambios en standard_price, pero eso creo que es problema del store=True del campo. Creo que tengo que quitarle el api.onchange y meter esta funcion en el write del producto para que tome los datos.

Reply:
No debería hacer falta pero bueno igual es por ser on_change y depends al mismo tiempo, no se. También, comprueba que la regla de la tarifa se basa en precio de coste, sino no dará precios correctos.

Reply:
Si, la regla se calcula con el precio de coste. Pero cuando llego a la funcion, si el precio de coste era 20 y lo cambio a 22, me llega 20 como precio de coste y el calculo del pvm me sale mal. Voy a comprobar quitando el on_change.

Reply:
No se proque se ejecuta la funcion, que esta dentro de product.product, cuando esta haciendo un write de product.template y me da un error de id.

Reply:
standard_price es un campo de product.template no de product.product por eso se te dispara desde ahí mueve tu campo de modelo y así ya debería funcionar..

Reply:
Ahora me salta este error al intentar hacer un update del context:
`ERROR - failed to write data to stream: <open file '<stdout>', mode 'w' at 0x7f73e0985150>`

Reply:
A lo mejor no funciona con la forma self.with_context(new_ctx), sino funciona así, recorres self.ids y dentro haces un self.with_context(new_ctx).browse(product_id)

Reply:
No, el problema no es el  self.with_context(new_ctx) sino al actualizar new_ctx con pricelist. Ahi es cuando me da el error que te he enviado

Reply:
pégame el código si eso a ver, porque ese error no me suena de nada

Reply:
![image](https://user-images.githubusercontent.com/22660449/35734314-011f59f0-0821-11e8-9589-e246e2f125c8.png)

Reply:
Es muy raro, el código lo veo bien, buscando por internet pone que puede ser por la codificación dle ipython, prueba a eliminar el ipdb a ver, o mira que lo esté ejecutando con python en lugar de ipython y que le fichero tenga el encoding en la primera linea

Reply:
El encoding esta, yo tambien he estado buscando y he encontrado lo mismo. Voy a quitar el ipdb y te cuento

Reply:
El ipdb no es. Como puedo saber si la ejecución la esta realizado python o ipython?

Reply:
No se como arrancas el Odoo, si lo haces desde el buidlout con ./start_openerp en teoría es python que será el que tines en la carpeta de sandbox/bin si lo arrancas con algo externo ya sea un editor u otra cosa tiene que ver que entorno de ejecución está usando, por lo general usan lo que tengas en la variable de entorno PYTHONPATH

Reply:
Lo arranco con el buildout, por lo que no deberia de tener problema en ese caso. No se si volver a hacer todo el proceso del buildout por si ha pasado algo.

Reply:
Es raro, puede probar a ejecutarlo Nadia antes a ver si también le falla, pero es bastante raro.
Es que no se me ocurre mucho más, otra forma de copiar un diccionario a parte de dict(ctx) es ctx.copy() puedes probar aunque en teoría son equivalentes, no creo que tampoco sea de esto ya que el error es como de escritura en un fichero físico no en memoria no parece estar relacionado con el contexto aunque falle en ese punto.

Reply:
Estoy haciendo la ejecucion en el equipo de Nadia, ahora te cuento

Reply:
A Nadia le funciona la ejecucion, pero por lo que hemos visto no lo almacena en BBDD. Algunas veces nos ha pasado esto con campos store=True, pero ahora despues de todos los problemas que he tenido no estoy seguro de que escriba bien

Reply:
con api.depends y store=True debería escribirlo en base de datos siempre, ¿mirastéis en la tabla product_template no?

Reply:
Si, miramos en esa tabla y encontramos el campo con valor 0. Esperamos a que se introdujeran los valores una vez ejecutamos bin/start_openerp

Reply:
Voy a subir los cambios, creo que podemos aceptarlos si dices que no ves nada mal en el código. EL problema de que el store no nos almacene la información nos ha pasado antes en local, por lo que creo que es un problema nuestro y no del codigo. Si no estas convencido del código yo esta tarde seguiré mirándolo mientras intento solucionar el problema del ipython.

Reply:
Metiendo las dos correcciones que comenté ahora lo veo bien, de todas forma slo dejo abierto por si haces alguna prueba más y detectas algo, lo mergearé en algún momento del finde.

Reply:
Perfecto, yo voy a intentar solucionar el problema con Python y si lo consigo, reviso la ejecución del código.